### PR TITLE
Add basic set of tests for ndctl userspace tool

### PR DESCRIPTION
**Pull request by harish-24**
Test adds basic set of tests for ndctl userspace tool which covers listing DIMMs, busses and enabling, disabling regions

Signed-off-by: Harish <harish@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by harish-24** [comment]
```
 (1/6) memory/ndctl.py:NdctlTest.test_bus_ids;upstream-2ae6: PASS (16.79 s)
 (2/6) memory/ndctl.py:NdctlTest.test_bus_ids;distro-fb52: PASS (0.09 s)
 (3/6) memory/ndctl.py:NdctlTest.test_dimms;upstream-2ae6: PASS (16.60 s)
 (4/6) memory/ndctl.py:NdctlTest.test_dimms;distro-fb52: PASS (0.08 s)
 (5/6) memory/ndctl.py:NdctlTest.test_regions;upstream-2ae6: PASS (16.66 s)
 (6/6) memory/ndctl.py:NdctlTest.test_regions;distro-fb52: PASS (0.18 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 50.66 s
```

File: memory/ndctl.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        Get the json of each provided options
+
+        return: By default returns entire detail of namespaces
+        """
+        try:
+            json_op = json.loads(process.system_output(
+                '%s list %s' % (self.binary, option), shell=True))
+        except ValueError:
+            json_op = []
+        if option:
```
[the remark this answers is not in the document]
Comment (harish-24) [reply memory/ndctl.py]: yes, but upcoming tests would require a way to get output without an option, so enabled it here.

File: memory/ndctl.py [hunk 2]
```diff
@@ -89,6 +89,73 @@ def enable_region(self, name='all'):
...
+    def create_namespace(self, region='', bus='', n_type='pmem', mode='fsdax',
+                         memmap='mem', name='', size='', uuid='',
+                         sector_size='', align='', force=False,
+                         autolabel=False):
+        """
+        Creates namespace with specified options
+        """
+        args = '-t %s -m %s ' % (n_type, mode)
+        if mode in ['fsdax', 'devdax']:
+            args = '%s -M %s' % (args, memmap)
```
Comment (narasimhan-v) [inline memory/ndctl.py]: `args += " %s" % memmap`

File: memory/ndctl.py [hunk 3]
```diff
@@ -89,6 +89,73 @@ def enable_region(self, name='all'):
...
+        if uuid:
+            args = '%s -u %s' % (args, uuid)
+        if sector_size:
+            args = '%s -l %s' % (args, sector_size)
+        if align:
+            args = '%s -a %s' % (args, align)
+        if force:
+            args = '%s -f' % args
+        if autolabel:
+            args = '%s -L' % args
```
Comment (narasimhan-v) [inline memory/ndctl.py]: see if you can optimise this snippet using some list or dict

**Comment by harish-24** [comment]
@narasimhan-v Pushed with the changes. Thanks.

**Comment by harish-24** [comment]
New tests log
```
 (4/5) memory/ndctl.py:NdctlTest.test_namespace;distro-fb52: PASS (0.53 s)
 (5/5) memory/ndctl.py:NdctlTest.test_namespace_modes;distro-fb52: FAIL: Namespace for ['raw'] mode failed! (0.67 s)
```

File: memory/ndctl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (narasimhan-v) [reply memory/ndctl.py]: ok

File: memory/ndctl.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                          shell=True, ignore_status=True):
+            self.fail('Namespace create command failed')
+
+    def destroy_namespace(self, namespace='all', region='', bus='',
+                          force=False):
+        """
+        Destroy namepsaces
+        """
+        args = namespace
+        if region:
```
Comment (narasimhan-v) [inline memory/ndctl.py]: could be similar to https://github.com/avocado-framework-tests/avocado-misc-tests/pull/1144/files#diff-bb3bf4b7fe06bf39ad148236e4b11a3dR116

Comment (narasimhan-v) [reply memory/ndctl.py]: in fact, these both can use the same dict

File: memory/ndctl.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        """
+        args_dict = {region: '-r', bus: '-b', name: '-n', size: '-s', uuid: '-u', sector_size: '-l', align: '-a'}
+        args = '-t %s -m %s ' % (n_type, mode)
+
+        for option in list(args_dict.keys()):
+            if option:
+                args += ' %s %s' % (args_dict[option], option)
+        if mode in ['fsdax', 'devdax']:
+            args += ' -M %s'% memmap
+        if force:
```
Comment (narasimhan-v) [inline memory/ndctl.py]: maybe -f and -L can be handled via another dict ? just a suggestion.

**Comment by narasimhan-v** [comment]
@harish-24 sorry for not looking thoroughly earlier, have some minor suggestions in line with earlier comments.

File: memory/ndctl.py (hunk 4 above)
Comment (harish-24) [reply memory/ndctl.py]: ok

**Comment by harish-24** [comment]
@narasimhan-v Pushed with suggested changes. Thanks.

File: memory/ndctl.py (hunk 4 above)
Comment (narasimhan-v) [reply memory/ndctl.py]: I meant initiating them at class scope.

File: memory/ndctl.py [hunk 6]
```diff
@@ -237,6 +237,37 @@ def test_namespace(self):
         namespaces = self.get_json('-N')
         self.log.info('Created namespace %s', namespaces)
 
+    def test_namespace_modes(self):
+        """
+        Create  different namespace types
+        """
+        failed = False
```
Comment (narasimhan-v) [inline memory/ndctl.py]: this can be handled looking into the length of failed_modes

Comment (harish-24) [reply memory/ndctl.py]: makes sense.

**Pull request merged**